File: src/DiagToolsValidator/DiagToolsValidator.Core/Configuration/CrossOSDACTest/CrossOSDACTestRunConfiguration.fs
namespace DiagToolsValidator.Core.Configuration

open System
open System.IO
open System.Collections.Generic
open System.Collections

open YamlDotNet.Serialization

open DiagToolsValidator.Core.Functionality

module CrossOSDACTestRunConfiguration =
    type SystemInformation() =
        member val EnvironmentVariables = new Dictionary<string, string>() with get, set

    type DotNetSetting() =
        member val SDKVersion: string = null with get, set
        member val DotNetRoot: string = null with get, set

    type DotNetDumpSetting() =
        member val ToolVersion: string = null with get, set
        member val ToolRoot: string = null with get, set
        member val Feed: string = null with get, set

    type TargetAppSetting() =
        member val BuildConfig: string = null with get, set
        member val OOMApp: DotNetApp.DotNetApp = new DotNetApp.DotNetApp() with get, set
        member val UHEApp: DotNetApp.DotNetApp = new DotNetApp.DotNetApp() with get, set

    type CrossOSDACTestBaseConfiguration() =
        member val TestBed: string = null with get, set
        member val SDKVersionList: string list = List.empty with get, set
        member val DotNetDump: DotNetDumpSetting = new DotNetDumpSetting() with get, set
        member val TargetApp: TargetAppSetting = new TargetAppSetting() with get, set
        member val SystemInfo: SystemInformation = new SystemInformation() with get, set
   
    
    type CrossOSDACTestRunConfiguration() =
        member val TestBed: string = null with get, set
        member val TestAssets: string = null with get, set
        member val DumpFolder: string = null with get, set
        member val AnalysisOutputFolder: string = null with get, set
        member val DotNet: DotNetSetting = new DotNetSetting() with get, set
        member val DotNetDump: DotNetDumpSetting = new DotNetDumpSetting() with get, set
        member val TargetApp: TargetAppSetting = new TargetAppSetting() with get, set
        member val SystemInfo: SystemInformation = new SystemInformation() with get, set

    type CrossOSDACTestConfiguration() =
        member val TestBed: string = null with get, set
        member val RunConfigurationList: CrossOSDACTestRunConfiguration list = List.empty with get, set

    [<AbstractClass;Sealed>]
    type CrossOSDACTestConfigurationGenerator private() =
        static let ParseConfigFile(path: string) = 
            let _deserializer = (new DeserializerBuilder()).IgnoreUnmatchedProperties().Build()
            let serializedConfiguration = File.ReadAllText(path)

            try
                let configuration = _deserializer.Deserialize<CrossOSDACTestBaseConfiguration>(serializedConfiguration);
                let parseExn = new exn("Fail to parse config file")
                if List.isEmpty configuration.SDKVersionList
                then
                    parseExn.Data.Add(nameof(CrossOSDACTestConfigurationGenerator), "Please specify .NET SDK version")
                    raise(parseExn)
                
                elif String.IsNullOrEmpty configuration.TestBed
                then
                    parseExn.Data.Add(nameof(CrossOSDACTestConfigurationGenerator), "Please specify testbed")
                    raise(parseExn)
                
                elif String.IsNullOrEmpty configuration.DotNetDump.ToolVersion
                then
                    parseExn.Data.Add(nameof(CrossOSDACTestConfigurationGenerator), "Please specify dotnet-dump version")
                    raise(parseExn)
                
                elif String.IsNullOrEmpty configuration.DotNetDump.Feed
                then
                    parseExn.Data.Add(nameof(CrossOSDACTestConfigurationGenerator), "Please specify dotnet-dump feed")
                    raise(parseExn)

                else 
                    configuration
            with ex -> 
                reraise()
         
        static member GenerateRunConfiguration(path: string) = 
            let baseConfig = ParseConfigFile path
            let aggregatedConfig = new CrossOSDACTestConfiguration()
            aggregatedConfig.TestBed <- baseConfig.TestBed
            aggregatedConfig.RunConfigurationList <-
                baseConfig.SDKVersionList
                |> List.map(
                    fun sdkVersion ->
                        let configuration = new CrossOSDACTestRunConfiguration()

                        configuration.TestBed <- baseConfig.TestBed
                        configuration.TestAssets <- Path.Combine(configuration.TestBed, "TestAssets")

                        configuration.DotNet.SDKVersion <- sdkVersion
                        configuration.DotNet.DotNetRoot <- Path.Combine(configuration.TestBed, $"dotnet-sdk-{sdkVersion}")

                        configuration.DumpFolder <- Path.Combine(configuration.TestAssets, "dump", configuration.DotNet.SDKVersion)
                        configuration.AnalysisOutputFolder <- Path.Combine(configuration.TestAssets, "analysisoutput", configuration.DotNet.SDKVersion)
                        
                        configuration.DotNetDump.ToolVersion <- baseConfig.DotNetDump.ToolVersion
                        configuration.DotNetDump.Feed <- baseConfig.DotNetDump.Feed
                        configuration.DotNetDump.ToolRoot <- Path.Combine(configuration.TestBed, $"dotnet-dump-sdk{sdkVersion}")
                    
                        for de: DictionaryEntry in Environment.GetEnvironmentVariables() |> Seq.cast<DictionaryEntry> do
                            configuration.SystemInfo.EnvironmentVariables[de.Key.ToString()] <- de.Value.ToString()
                        if DotNet.CurrentRID.Contains("win")
                        then 
                            configuration.SystemInfo.EnvironmentVariables["Path"] <- 
                                Environment.GetEnvironmentVariable("Path") + $";{configuration.DotNet.DotNetRoot}"
                        else
                            configuration.SystemInfo.EnvironmentVariables["Path"] <- 
                                Environment.GetEnvironmentVariable("Path") + $":{configuration.DotNet.DotNetRoot}"
                        configuration.SystemInfo.EnvironmentVariables["DOTNET_ROOT"] <- configuration.DotNet.DotNetRoot

                        let targetAppsRoot = Path.Combine(configuration.TestAssets, "TargetApps", configuration.DotNet.SDKVersion)
                        configuration.TargetApp.BuildConfig <- baseConfig.TargetApp.BuildConfig
                        configuration.TargetApp.OOMApp <- new DotNetApp.DotNetApp(configuration.SystemInfo.EnvironmentVariables,
                                                                                  "console",
                                                                                  Path.Combine(targetAppsRoot, $"oom-sdk{sdkVersion}-{DotNet.CurrentRID}"))
                        configuration.TargetApp.UHEApp <- new DotNetApp.DotNetApp(configuration.SystemInfo.EnvironmentVariables,
                                                                                  "console",
                                                                                  Path.Combine(targetAppsRoot, $"uhe-sdk{sdkVersion}-{DotNet.CurrentRID}"))
                        
                        configuration)

            aggregatedConfig
            

File: src/DiagToolsValidationRunner/DiagToolsValidationRunner.Core/Configuration/CrossOSDACTestRunConfiguration.cs
using System.Collections;

using DiagToolsValidationRunner.Core.Functionality;

using YamlDotNet.Serialization;

namespace DiagToolsValidationRunner.Core.Configuration.CrossOSDACTest
{
    public class TargetAppSetting : BaseTargetAppSetting
    {
        public DotNetApp OOM = new();
        public DotNetApp UHE = new();
    }

    public class TestSetting : BaseTestSetting
    {
        public string DumpFolder = String.Empty;
        public string AnalysisOutputFolder = String.Empty;
    }

    public class CrossOSDACTestRunConfiguration
    {
        public required TestSetting Test;
        public required DotNetSDKSetting SDKSetting;
        public required DotNetToolSetting DotNetDumpSetting;
        public required TargetAppSetting AppSetting;
        public required BaseSystemInformation SysInfo;
    }

    public class CrossOSDACTestConfiguration
    {
        public required BaseTestSetting Test;
        public required List<string> SDKVersionList;
        public required DotNetToolSetting DotNetDumpSetting;
        public required TargetAppSetting AppSetting;
        public List<CrossOSDACTestRunConfiguration> CrossOSDACTestRunConfigurationList = new();
    }

    public static class CrossOSDACTestConfigurationGenerator
    {
        private static IDeserializer _deserializer = (new DeserializerBuilder()).IgnoreUnmatchedProperties().Build();

        private static CrossOSDACTestConfiguration ParseConfigFile(string configFile)
        {
            string serializedConfiguration = File.ReadAllText(configFile);
            CrossOSDACTestConfiguration baseConfiguration =
                _deserializer.Deserialize<CrossOSDACTestConfiguration>(configFile);

            if (string.IsNullOrEmpty(baseConfiguration.Test.TestBed))
            {
                throw new Exception(
                    $"{nameof(CrossOSDACTestConfigurationGenerator)}: Please specify testbed");
            }

            else if (baseConfiguration.SDKVersionList.Count == 0)
            {
                throw new Exception(
                    $"{nameof(CrossOSDACTestConfigurationGenerator)}: Please specify at least one SDK Version");
            }

            else if (string.IsNullOrEmpty(baseConfiguration.DotNetDumpSetting.Version))
            {
                throw new Exception(
                    $"{nameof(CrossOSDACTestConfigurationGenerator)}: Please specify tool version");
            }

            else if (string.IsNullOrEmpty(baseConfiguration.DotNetDumpSetting.Feed))
            {
                throw new Exception(
                    $"{nameof(CrossOSDACTestConfigurationGenerator)}: Please specify tool install feed");
            }

            else if (!(new List<string> { "Debug", "Release" }.Contains(baseConfiguration.AppSetting.BuildConfig)))
            {
                throw new Exception(
                    $"{nameof(CrossOSDACTestConfigurationGenerator)}: Please specify valid build config");
            }

            return baseConfiguration;
        }

        private static CrossOSDACTestRunConfiguration CreateRunConfig(string testbed,
                                                                      string testResultFolder,
                                                                      string testName,
                                                                      string dumpFolder,
                                                                      string analysisOutputFolder,
                                                                      string dotNetDumpVersion,
                                                                      string dotNetDumpFeed,
                                                                      string dotNetDumpRoot,
                                                                      string SDKVersion,
                                                                      string dotNetRoot,
                                                                      Dictionary<string, string> env,
                                                                      string buildConfig,
                                                                      string oomAppRoot,
                                                                      string uheAppRoot)
        {
            return new()
            {
                Test = new()
                {
                    TestBed = testbed,
                    TestResultFolder = testResultFolder,
                    TestName = testName,
                    DumpFolder = dumpFolder,
                    AnalysisOutputFolder = analysisOutputFolder
                },
                DotNetDumpSetting = new()
                {
                    Version = dotNetDumpVersion,
                    Feed = dotNetDumpFeed,
                    ToolRoot = dotNetDumpRoot
                },
                SDKSetting = new()
                {
                    Version = SDKVersion,
                    DotNetRoot = dotNetRoot,
                },
                SysInfo = new()
                {
                    EnvironmentVariables = new(env)
                },
                AppSetting = new()
                {
                    BuildConfig = buildConfig,
                    OOM = new(env, "console", oomAppRoot),
                    UHE = new(env, "console", uheAppRoot)
                }
            };
        }

        public static CrossOSDACTestConfiguration GenerateConfiguration(string configFile)
        {
            CrossOSDACTestConfiguration configuration = ParseConfigFile(configFile);
            string testResultFolder = Path.Combine(configuration.Test.TestBed, "TestResult");


            configuration.CrossOSDACTestRunConfigurationList = new();
            foreach (string SDKVersion in configuration.SDKVersionList)
            {
                if (OperatingSystem.IsWindows())
                {
                    List<string> winRIDList = new() { "win-x86", "win-x64" };
                    foreach (var rid in winRIDList)
                    {
                        string testName = $"CrossOSDAC-SDK{SDKVersion}-{rid}";
                        string dumpFolder = Path.Combine(testResultFolder, $"dumps-sdk{SDKVersion}");
                        string analysisOutputFolder = Path.Combine(testResultFolder, $"analysis-sdk{SDKVersion}");

                        string dotNetRoot = Path.Combine(configuration.Test.TestBed, $"DotNetSDK-{SDKVersion}-{rid}");

                        string dotNetDumpRoot = Path.Combine(configuration.Test.TestBed, $"dotnet-dump-{SDKVersion}-{rid}");

                        Dictionary<string, string> env = new();
                        foreach (DictionaryEntry de in Environment.GetEnvironmentVariables())
                        {
                            env[de!.Key!.ToString()!] = de!.Value!.ToString()!;
                        }
                        env["DOTNET_ROOT"] = dotNetRoot;

                        string targetAppsRoot = Path.Combine(testResultFolder, "TargetApps");
                        string oomAppRoot = Path.Combine(targetAppsRoot, $"oom-sdk{SDKVersion}");
                        string uheAppRoot = Path.Combine(targetAppsRoot, $"uhe-sdk{SDKVersion}");

                        CrossOSDACTestRunConfiguration runConfig = CreateRunConfig(configuration.Test.TestBed,
                                                                                   testResultFolder,
                                                                                   testName,
                                                                                   dumpFolder,
                                                                                   analysisOutputFolder,
                                                                                   configuration.DotNetDumpSetting.Version,
                                                                                   configuration.DotNetDumpSetting.Feed,
                                                                                   dotNetDumpRoot,
                                                                                   SDKVersion,
                                                                                   dotNetRoot,
                                                                                   env,
                                                                                   configuration.AppSetting.BuildConfig,
                                                                                   oomAppRoot,
                                                                                   uheAppRoot);

                        configuration.CrossOSDACTestRunConfigurationList.Add(runConfig);
                    }
                }
                else if (OperatingSystem.IsLinux())
                {
                    string testName = $"CrossOSDAC-SDK{SDKVersion}";
                    string dumpFolder = Path.Combine(testResultFolder, $"dumps-sdk{SDKVersion}");
                    string analysisOutputFolder = Path.Combine(testResultFolder, $"analysis-sdk{SDKVersion}");
                    string dotNetRoot = Path.Combine(configuration.Test.TestBed, $"DotNetSDK-{SDKVersion}");
                    string dotNetDumpRoot = Path.Combine(configuration.Test.TestBed, $"dotnet-dump-{SDKVersion}");

                    Dictionary<string, string> env = new();
                    foreach (DictionaryEntry de in Environment.GetEnvironmentVariables())
                    {
                        env[de!.Key!.ToString()!] = de!.Value!.ToString()!;
                    }
                    env["DOTNET_ROOT"] = dotNetRoot;

                    string targetAppsRoot = Path.Combine(testResultFolder, "TargetApps");
                    string oomAppRoot = Path.Combine(targetAppsRoot, $"oom-sdk{SDKVersion}");
                    string uheAppRoot = Path.Combine(targetAppsRoot, $"uhe-sdk{SDKVersion}");

                    CrossOSDACTestRunConfiguration runConfig = CreateRunConfig(configuration.Test.TestBed,
                                                                               testResultFolder,
                                                                               testName,
                                                                               dumpFolder,
                                                                               analysisOutputFolder,
                                                                               configuration.DotNetDumpSetting.Version,
                                                                               configuration.DotNetDumpSetting.Feed,
                                                                               dotNetDumpRoot,
                                                                               SDKVersion,
                                                                               dotNetRoot,
                                                                               env,
                                                                               configuration.AppSetting.BuildConfig,
                                                                               oomAppRoot,
                                                                               uheAppRoot);

                    configuration.CrossOSDACTestRunConfigurationList.Add(runConfig);
                }
                else
                {
                    throw new Exception($"{nameof(CrossOSDACTestConfigurationGenerator)}: Unsupported OS");
                }
            }

            return configuration;
        }
    }
}
